### DISCUSSION TITLE
Request for collaboration

### DISCUSSION THREAD

**Discussion by kotfu** [hamdanal/rich-argparse#61] [General]
Hey, I'm one of the maintainers of [cmd2](https://github.com/python-cmd2/cmd2), and we are exploring ways we can either incorporate [rich](https://github.com/Textualize/rich) into our library as a dependency, and/or make it easier for users of `cmd2` to utilize `rich` in their own applications. Our issue capturing the discussion so far is at https://github.com/python-cmd2/cmd2/issues/1251.

In my early exploration of how best to do this in cmd2, I found this excellent project, which solves an important part of the problem for us. I have identified an initial list of changes/improvements to rich-argparse which would really benefit our work. Here's a brief list:

- we would like more styles and more granular styles for different elements in the help/usage output
- the current implementation creates it's own `rich.console` objects, we would be interested in exploring changes to rich-argparse which could make it easier for a user/subclasser of `RichHelpFormatter()` to provide their own `rich.console` object (which I get will be tricky or maybe impossible since `argparse.ArgumentParser` instantiates this class.
- the current implementation of `RichHelpFormatter.format_help()` uses `rich.console.capture` which seems to create some formatting problems if you have "embedded" `rich.console` objects, ie one console object processing the output of `format_help()` while another console is capturing inside of `format_help()`. Not sure yet if this is a `rich-argparse` problem or a `rich` problem.

If you are open to collaboration/discussion/pull requests of our changes/enhancements, we would love to work with you. That would be our preferred approach if that works for you. If not, that's fine too, the MIT license for this project is compatible with the cmd2 license and we could fork your work to make it work better for our use case.

Are you open to collaboration, discussion, and pull requests from me?

**Comment by hamdanal** [comment]
Hi, I'd be happy to make changes that'll make it easier for you to use `rich_argparse` as long as they fit into this project and they don't break current users. I don't know what cmd2 does exactly or how it uses rich so please answer my questions.

> * we would like more styles and more granular styles for different elements in the help/usage output

Could you clarify what styles are you thinking of? perhaps give an example.

> * the current implementation creates it's own `rich.console` objects, we would be interested in exploring changes to rich-argparse which could make it easier for a user/subclasser of `RichHelpFormatter()` to provide their own `rich.console` object (which I get will be tricky or maybe impossible since `argparse.ArgumentParser` instantiates this class.

Could you please explain the use case for a custom console? The console in `RichHelpFormatter` is only used to produce the correct string output of the help text. The only reason a console is used is because rich objects do not offer an interface to get their raw text as pointed out by rich creator here https://github.com/Textualize/rich/issues/1179#issuecomment-824308067. Users are not supposed to print anything using this console that is not produced by the formatter.

> * the current implementation of `RichHelpFormatter.format_help()` uses `rich.console.capture` which seems to create some formatting problems if you have "embedded" `rich.console` objects, ie one console object processing the output of `format_help()` while another console is capturing inside of `format_help()`. Not sure yet if this is a `rich-argparse` problem or a rich problem.

The output of `RichHelpFormatter.format_help` is a raw string with *ansi* codes produced by `console.print`. It is supposed to be written as is to the output file (in the argparse case this is stdout). I am not sure how exactly you are using this output or why you'll want to print it with a console but if you want to get a "rich object" that you can manipulate from the help output, you can use [`Text.from_ansi`](https://rich.readthedocs.io/en/stable/reference/text.html#rich.text.Text.from_ansi) to do that.

> Are you open to collaboration, discussion, and pull requests from me?

Sure thing. You are more that welcome to discuss or contribute pull requests. I cannot promise that I can accept all your requests so a discussion before you spend time implementing a PR would be appreciated.

**Comment by kotfu** [comment]
That all sounds great. I'll open a new issue for each proposal for discussion and provide more detail. You can use that discussion to determine if the proposal fits with your vision for the project and if so, I'm happy to work on pull requests which you can then review to ensure the implementation is satisfactory, or if you'd like changes to be made before merging (or rejecting all together, as you see fit).

**Comment by hamdanal** [comment]
@kotfu after #54 and #60, I feel there is something about your use case that I don't understand yet. Could you please state here why exactly you need to print the text produced by `RichHelpFormatter.format_help` using a bare `console.print` call instead of the other valid methods that I suggested in those issues? I would like to understand more how you are using this library so that I can help you find a proper solution. Perhaps you could include a full example of how your program is constructed and how it integrates rich-argparse.
